Zend/zend_execute.c

```diff
@@ -1685,7 +1693,12 @@ static void zend_incdec_typed_prop(zend_property_info *prop_info, zval *var_ptr,
 		decrement_function(var_ptr);
 	}
 
-	if (UNEXPECTED(Z_TYPE_P(var_ptr) == IS_DOUBLE) && Z_TYPE_P(copy) == IS_LONG) {
+	if (UNEXPECTED(prop_info->flags & ZEND_ACC_FINAL)) {
```

Comment:
You can move this check earlier and save some of the copy juggling. The other checks occur after because they compare the old and new vals.